### PR TITLE
Add dataset for running with precomputed latents from multiple captions

### PR DESCRIPTION
**Pull request by coryMosaicML**
An extension of [151](https://github.com/mosaicml/diffusion/pull/151) that can now be closed.

### CODE REVIEW

File: diffusion/datasets/image_caption_latents.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+from typing import Callable, Dict, List, Optional, Sequence, Tuple, Union
+
+import numpy as np
+import torch
+from PIL import Image
+from streaming import Stream, StreamingDataset
+from torch.utils.data import DataLoader
+from torchvision import transforms
+
+from diffusion.datasets.laion.transforms import LargestCenterSquare, RandomCropAspectRatioTransorm, RandomCropSquare
```
Comment (gupta-abhay) [inline diffusion/datasets/image_caption_latents.py:17]: nit: missing F in `RandomCropAspectRatioTransorm` ?

File: diffusion/datasets/image_caption_latents.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        # https://github.com/Stability-AI/generative-models/blob/477d8b9a7730d9b2e92b326a770c0420d00308c9/sgm/modules/encoders/modules.py#L151-L160
+        if torch.rand(1) < self.microcond_drop_prob:
+            out['cond_crops_coords_top_left'] = out['cond_crops_coords_top_left'] * 0
+        if torch.rand(1) < self.microcond_drop_prob:
+            out['cond_original_size'] = out['cond_original_size'] * 0
+        if torch.rand(1) < self.microcond_drop_prob:
+            out['cond_target_size'] = out['cond_target_size'] * 0
+
+        # Randomly select a caption according to the selection probabilities
+        caption_key = np.random.choice(self.caption_keys, p=self.caption_selection_probs)
```
Comment (gupta-abhay) [inline diffusion/datasets/image_caption_latents.py:119]: ques: why use `torch.rand` in one place vs `np.random` in another - are the seeds for torch and numpy rngs set differently and hence it ensures that the two selections are exclusive from each other?

File: diffusion/datasets/image_caption_latents.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            latent_shape = self.text_latent_shapes[i]
+            attention_key = f'{caption_key}_{self.attention_mask_keys[i]}'
+
+            if torch.rand(1) < self.caption_drop_prob:
+                out[self.text_latent_keys[i]] = torch.zeros(latent_shape, dtype=torch.float16)
+                out[self.attention_mask_keys[i]] = torch.zeros(latent_shape[0])
+                if 'CLIP_LATENTS' in latent_key:
+                    out['CLIP_POOLED'] = torch.zeros(latent_shape[1])
+            else:
+                text_latent = np.frombuffer(sample[latent_key], dtype=np.float16).copy()
```
Comment (gupta-abhay) [inline diffusion/datasets/image_caption_latents.py:132]: confirming that the dtype for the latents are valid for training - ie, model expects tensors in float16 as well?

**Review by gupta-abhay — APPROVED** [review]
LGTM! Left some minor comments.

File: diffusion/datasets/image_caption_latents.py (hunk 2 above)
Comment (coryMosaicML) [reply diffusion/datasets/image_caption_latents.py:119]: There is no real reason to do this AFAIK. This is just consistent with our other datasets but should probably be fixed

File: diffusion/datasets/image_caption_latents.py (hunk 1 above)
Comment (coryMosaicML) [reply diffusion/datasets/image_caption_latents.py]: Should be fixed now!

File: diffusion/datasets/image_caption_latents.py (hunk 3 above)
Comment (coryMosaicML) [reply diffusion/datasets/image_caption_latents.py:132]: We are planning on this, but perhaps it should be configurable

Comment (coryMosaicML) [reply diffusion/datasets/image_caption_latents.py]: Changed it to be configurable!

**Pull request merged**